File: Data_Structures/Tree/Binary_Search_Tree/Validate_BST/Validate_BST.java
import java.util.*;

class Node {
//class describing structure of the tree node
  int value;
  Node left;
  Node right;
  Node(int value) { 
    this.value = value; 
    left = right = null;
    }
 }
 
 class Solution {
 //Recursive functions to determine whether given tree is valid BST or not
    public boolean isValidBST(Node root) {
        return helper(root,null,null);
    }
    public boolean helper(Node root, Integer small, Integer large) 
    {
        if(root==null)
            return true;
        if(small!=null && root.val<=small)
            return false;
        if(large!=null && root.val>=large)
            return false;
        return helper(root.left,small,root.val) && helper(root.right,root.val,large);
    }
}

public class Validate_BST
{
//main class
  public static void main(String[] args){
        Scanner scanner = new Scanner(System.in);
        Solution object = new Solution();
        Node TreeRoot1 = new Node(10);
        TreeRoot1.left = new Node(5);
        TreeRoot1.right = new Node(12);
        TreeRoot1.left.left = new Node(4);
        TreeRoot1.left.right = new Node(6);
        boolean answer1 = object.isValidBST(TreeRoot1);
        if(answer1==true)
          System.out.println("Tree is a valid BST");
        else
          System.out.println("Tree is not a valid BST");

        Node TreeRoot2 = new Node(2);
        TreeRoot2.left = new Node(6);
        TreeRoot2.right = new Node(1);
        TreeRoot2.left.right = new Node(5);
        TreeRoot2.right.left = new Node(8);
        TreeRoot2.right.right = new Node(11);
        boolean answer2 = object.isValidBST(TreeRoot2);
        if(answer2==true)
          System.out.println("Tree is a valid BST");
        else
          System.out.println("Tree is not a valid BST");

  }
}

File: Algorithms/Tree-Algorithm/Validate_BST/Validate_BST.cs
using System;

internal class Node
{
    public int value;
    public Node left, right;

    public Node(int item)
    {
        value = item;
        left = right = null;
    }
}

public class Validate_BST
{
    static bool IsBST(
        Node node,
        int currentMinimum = Int32.MinValue,
        int currentMaximum = Int32.MaxValue)
    {
        if (node == null)
        {
            return true;
        }

        if (node.value < currentMinimum || node.value > currentMaximum)
        {
            return false;
        }

        return (IsBST(node.left, currentMinimum, node.value - 1) &&
                IsBST(node.right, node.value + 1, currentMaximum));
    }

    public static void Main(string[] args)
    {
        Node treeRoot1 = new Node(7);
        treeRoot1.left = new Node(5);
        treeRoot1.right = new Node(8);
        treeRoot1.left.left = new Node(4);
        treeRoot1.left.right = new Node(6);
        Console.WriteLine(IsBST(treeRoot1) ? "This tree is a BST" : "This tree is not a BST");

        Node treeRoot2 = new Node(2);
        treeRoot2.left = new Node(6);
        treeRoot2.left.right = new Node(5);
        treeRoot2.right = new Node(1);
        treeRoot2.right.left = new Node(8);
        treeRoot2.right.right = new Node(10);
        Console.WriteLine(IsBST(treeRoot2) ? "This tree is a BST" : "This tree is not a BST");
    }
}
